File: officialdrops/github_add_comment.ts
/**
 * github_add_comment — official scripted connector (replaces the native Go
 * drop). Posts a Markdown comment to a GitHub issue or PR (same number space).
 */

const GITHUB_API_BASE = "https://api.github.com";

export default {
  manifest: {
    id: "github_add_comment",
    version: "2.0.0",
    label: "GitHub add comment",
    summary: "Post a Markdown comment to a GitHub issue or pull request via the REST API.",
    description:
      "Add a comment to a GitHub issue or pull request. Works for both — GitHub treats them as the same under the hood. The body supports Markdown and can come from the 'body' input or params.",
    integration: "GitHub",
    category: "network",
    icon: "git-branch",
    brandLogo: "/brands/github.svg",
    color: "#24292f",
    tags: ["github", "issue", "pr", "comment"],
    requiresConnections: [{ kind: "oauth", name: "github", note: "GitHub OAuth." }],
    inputs: [{ port: "body", label: "Comment body (overrides params.body; Markdown)" }],
    outputs: [{ port: "meta", label: "Created comment metadata", mime: ["application/json"] }],
    idempotent: false,
    retryPolicy: "exponential_backoff",
    paramsSchema: {
      type: "object",
      properties: {
        account: { type: "string", default: "default" },
        token: { type: "string" },
        owner: { type: "string" },
        repo: { type: "string" },
        issue_number: { type: "integer", description: "Issue OR PR number (shared number space)." },
        body: { type: "string", description: "Comment body. Overridden by the 'body' input port." },
        timeout_ms: { type: "integer", default: 15000, minimum: 1 },
      },
      required: ["owner", "repo", "issue_number"],
    },
    examples: [
      { title: "Acknowledge a triage issue", params: { owner: "example", repo: "widgets", issue_number: 142, body: "Thanks — reproduced on main.", token: "${secret:GITHUB_TOKEN}" } },
    ],
  },

  async run(ctx: any) {
    const p = ctx.params || {};
    const owner = String(p.owner || "").trim();
    const repo = String(p.repo || "").trim();
    const issueNumber = Number(p.issue_number) || 0;
    if (!owner || !repo) throw new DropError("bad_param", "'owner' and 'repo' are required");
    if (issueNumber <= 0) throw new DropError("bad_param", "issue_number must be a positive integer");

    let token = String(p.token || "").trim();
    if (!token) {
      if (!ctx.auth) throw new DropError("auth", "no token supplied and OAuth is not configured");
      token = await ctx.auth.token("github", p.account || "default");
    }

    let body = String(p.body || "");
    if (ctx.inputs.has("body")) {
      const ref = ctx.inputs.ref("body");
      if (typeof ref.value === "string") body = ref.value;
      else if (ref.path) body = await ctx.files.readText(ref.path);
      else if (ref.value !== undefined) body = "```json\n" + JSON.stringify(ref.value, null, 2) + "\n```";
    }
    if (!body) throw new DropError("bad_input", "comment body is empty");

    const res = await ctx.fetch(
      `${GITHUB_API_BASE}/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/issues/${issueNumber}/comments`,
      {
        method: "POST",
        headers: githubHeaders(token),
        body: JSON.stringify({ body }),
        timeoutMs: Number(p.timeout_ms) || 15000,
      },
    );
    if (!res.ok) throw new DropError("github_error", githubError(res.status, await res.text()));
    const c: any = await res.json();
    return { meta: { id: c.id || 0, node_id: c.node_id || "", html_url: c.html_url || "" } };
  },
};

function githubHeaders(token: string): Record<string, string> {
  return {
    Authorization: `Bearer ${token}`,
    Accept: "application/vnd.github+json",
    "X-GitHub-Api-Version": "2022-11-28",
    "Content-Type": "application/json",
  };
}

function githubError(status: number, text: string): string {
  try {
    const j = JSON.parse(text);
    if (j && j.message) {
      if (Array.isArray(j.errors) && j.errors.length) {
        const e = j.errors[0];
        if (e.message) return `${j.message}: ${e.message}`;
        if (e.field) return `${j.message}: field "${e.field}" (${e.code})`;
      }
      return j.message;
    }
  } catch (_e) {
    // not JSON
  }
  return `GitHub returned ${status}: ${text.slice(0, 512)}`;
}

File: drops/github/github_add_comment.go
// SPDX-FileCopyrightText: 2026 Angels' Ware
// SPDX-License-Identifier: AGPL-3.0-or-later

package github

import (
	"context"
	"encoding/json"
	"net/url"
	"strconv"

	"git.sr.ht/~klahr/dazyflow/core"
	"git.sr.ht/~klahr/dazyflow/drops/internal/params"
	"git.sr.ht/~klahr/dazyflow/engine"
)

func init() {
	engine.Register(engine.NativeDrop{
		Manifest: core.Manifest{
			ID:          "github_add_comment",
			Version:     "1.0",
			Label:       "GitHub",
			Subtitle:    "Add comment",
			Summary:     "Add a comment to a GitHub issue or pull request.",
			Description: "Comment on a GitHub issue or PR (they share a number space). The comment can be typed on the step or connected in from another step (the input overrides the typed value); Markdown works. Outputs a link to the posted comment.",
			Integration: "GitHub",
			Category:    "network",
			Icon:        "git-branch",
			BrandLogo:   "/brands/github.svg",
			Color:       "#24292f",
			Provider:    "internal",
			Tags:        []string{"github", "comment", "issue", "pr"},
			Examples: []core.ParamsExample{
				{Title: "Acknowledge a triage issue", Params: json.RawMessage(`{"owner":"example","repo":"widgets","issue_number":142,"body":"Thanks — reproduced on main.","token":"${secret.GITHUB_TOKEN}"}`)},
			},
			RequiresConnections: []core.ConnectionRequirement{
				{Kind: "oauth", Name: "github", Note: "GitHub OAuth — repo scope (issues:write)."},
			},
			ExecutionModel: core.ExecutionBatch,
			ProcessModel:   core.ProcessLongLived,
			Inputs: []core.Port{
				// Named after its param so the card shows an inline editable
				// box (Unreal-style); a wired value overrides the typed one.
				{Port: "body", Label: "Comment"},
			},
			Outputs: []core.Port{
				// Only the friendly scalar is a pin; the full comment metadata
				// (id, node_id, …) is still EMITTED under "meta" so run records
				// keep it for debugging — it's just not a pin.
				{Port: "comment_url", Label: "Comment link", MIME: []string{"text/plain"}},
				{Port: "meta", Label: "Details", MIME: []string{"application/json"}},
			},
			ParamsSchema: json.RawMessage(`{
				"type":"object",
				"properties":{
					"account":{"type":"string","default":"default"},
					"token":{"type":"string"},
					"owner":{"type":"string","title":"Repo owner","description":"The username or organization the repo lives under."},
					"repo":{"type":"string","title":"Repo name","description":"The repo's name, without the owner part."},
					"issue_number":{"type":"integer","title":"Issue or PR number","description":"The number from the issue/PR link — issues and PRs share one number space."},
					"body":{"type":"string","title":"Comment","description":"What to say (Markdown works). Overridden by the 'Comment' input."},
					"timeout_ms":{"type":"integer","default":15000,"minimum":1,"description":"Hard deadline for the request, in milliseconds."}
				},
				"required":["owner","repo","issue_number"]
			}`),
			Idempotent:  false,
			RetryPolicy: core.RetryExponentialBackoff,
			// A non-idempotent external write: opt into engine-side dedupe so an
			// expired-lease reclaim or crash recovery replays the recorded result
			// instead of firing the write a second time. Matches the other
			// send-style drops (discord/gmail/sheets/twilio/klarna/nshift/elks).
			DedupeWrites: true,
		},
		Execute: executeGitHubAddComment,
	})
}

func executeGitHubAddComment(ctx context.Context, job core.Job, _ chan<- core.Progress) (core.Result, error) {
	owner, _ := params.StringOpt(job.Params, "owner")
	repo, _ := params.StringOpt(job.Params, "repo")
	if owner == "" || repo == "" {
		return params.Err(job, "bad_param", "'owner' and 'repo' are required"), nil
	}
	issueNumber := params.IntDefault(job.Params, "issue_number", 0)
	if issueNumber <= 0 {
		return params.Err(job, "bad_param", "issue_number must be a positive integer"), nil
	}
	token, err := resolveToken(ctx, job)
	if err != nil {
		return params.Err(job, "auth", err.Error()), nil
	}
	body := resolveBody(job)
	if body == "" {
		return params.Err(job, "bad_input", "comment body is empty"), nil
	}
	raw, _ := json.Marshal(map[string]any{"body": body})

	endpoint := currentHTTPBase() + "/repos/" + url.PathEscape(owner) + "/" + url.PathEscape(repo) + "/issues/" + strconv.Itoa(issueNumber) + "/comments"
	// Posting a comment is a non-idempotent POST and this drop is a
	// terminal leaf, so the engine auto-retries it — pass the job's stable
	// Idempotency-Key (GitHub honors the header) so a replay dedupes
	// server-side instead of posting a duplicate comment.
	status, respBody, err := githubDoIdem(ctx, "POST", endpoint, token, raw, params.IntDefault(job.Params, "timeout_ms", 15000), job.IdempotencyKey())
	if err != nil {
		return params.Err(job, "github_http_error", err.Error()), nil
	}
	if status < 200 || status >= 300 {
		return params.Err(job, "github_error", extractGitHubError(respBody)), nil
	}

	var c struct {
		ID      int64  `json:"id"`
		NodeID  string `json:"node_id"`
		HTMLURL string `json:"html_url"`
	}
	_ = json.Unmarshal(respBody, &c)
	return core.Result{
		JobID:  job.ID,
		Status: core.StatusOK,
		Output: map[string]core.Ref{
			"comment_url": {MIME: "text/plain", Inline: c.HTMLURL},
			"meta": {MIME: "application/json", Inline: map[string]any{
				"id": c.ID, "node_id": c.NodeID, "html_url": c.HTMLURL,
			}},
		},
	}, nil
}
